modxcms/revolution
# Fix and Enhance Usergroup ACL Permissions Grids and Filtering

### What does it do?
Made a number of changes and additions in the processors and grid JS files to correct issues with the grids’ grouping functionality and more accurately build the grid filters depending on the data shown. More specifically:

1. Replaced the column model renderers (used to render links on column data) with a `templatecolumn` specification (generated by a new method in the grids base class), as using renderers on grouping grids interferes with the grouping functionality (because a renderer is an interceptor method in Ext).
2. Moved menu specifications (grid context menus) out of processors and into the grid js for all sections (context was already this way). The primary reason being that some events (such as `afterRemoveRow`) were not firing when set up this way; monitoring these events is necessary for maintaining accurate filter options when making changes to a grid's data. Also, there's no reason to specify js in php unless the item in question will change on the fly and needs setup data from the processor.
3. Grid view, including the bottom toolbar, now correctly updates when a row is deleted.
4. Created new methods to update dependent filters (when two or more filters are present) based upon the selected value of any given filter.
5. Fixed some filters that were missing options due to an incorrect specification of their `baseParams.group` property.

### Why is it needed?
Grouping in these grids was not working correctly. Also, providing filter options where no relevant data is available in a given grid is a UX problem that needs to be fixed not only here, but across all grids with filters. 

### How to test
1. Ensure modx and browser caches are fully cleared.
2. Create a number of policies in each area (if you don't already have some set up) and play with the grouping and filtering of items in each grid. (These are the five Access Permissions grids — Context, Resource Groups, Element Categories, Media Sources, and Namespaces.) Verify that filters only show relevant options depending on the grid data and other filters you've made selections on.
3. Verify that other operations (editing and deleting) affect filter options as expected.

### Related issue(s)/PR(s)
Resolves #16239.

### Special Note
Because these updates involve a lot of changes, I didn't want to execute them for _all_ relevant grids at once. This PR completes a logical section and, depending on any requested changes, other PRs will follow what I've done here for other grids.

### Before and After Examples
Currently, in MODx 3.0.3-dev, note each of the following problems (in order of appearance in the video):

1. Each ACL area grid does not group properly; where an area's rows should be grouped under a common heading, they are not. To see this clearly, pause each time I click on each area and note where grouping should be occurring (for example, in Contexts, the three assigned the _Editors - 1000_ Role should be grouped under one row heading for that Role).
2. [0:19] Next, notice how the grid filters show irrelevant options where they should only show those present in the grid data. The _Filter by Context_ combo shows all Contexts instead of those present (dupctx1, dupctx2, dev2, and www2). The same goes for the _Filter by Policy_ combo. [0:42] Additionally, the Filter by Policy combo is mis-configured in places; here the problem in the Contexts area is illustrated. When adding a new Context, you are given the full Policies list in the Create window, but the filter combo shows an incomplete subset. [1:02] Lastly, when attempting to create additional policies for the same Context, an error occurs due to an incorrect validation for a unique context key. 
3. [1:32] When deleting a row, see how the count is not updated in the lower right grid status area.
4. [1:50] Next, when filtering, a choice in one combo should update the others to show only entries relevant to the current state of the query. They do not currently react to changes as one would expect.

https://user-images.githubusercontent.com/689075/209720912-944ca7bb-2d91-4aae-9bdd-f0b5b7d71b18.mov

Now, note how each problem is fixed (action taken in same order as above):

https://user-images.githubusercontent.com/689075/209720965-f3b53940-7554-4db1-a695-86bc2b25c8b7.mov
